### DISCUSSION TITLE
Refactor template to expose event callback to navigate

### DISCUSSION THREAD

**Discussion by kaungkhantsoe** [nimblehq/android-templates#593] [RFCs]
## Issue
- Currently, our template uses a `navigator` from the BaseViewModel to navigate around, which is not necessary

## Solution
- We need to refactor it to follow the compose's [Unidirection Data Flow](https://developer.android.com/develop/ui/compose/architecture#udf) as we do in many of our client projects
- To do so, we will refactor the whole project to navigate through event callbacks and remove `navigator` from the Screen level and ViewModel level
- We will also apply some useful custom extensions from our client's project, like `NavController.navigate`

## Who benefits?
Android developers

## What's next?

**Comment by hoangnguyen92dn** [comment]
> Currently, our template uses a navigator from the BaseViewModel to navigate around, which is not necessary

This approach has pros/cons and the team already discussed about this. Mentioning "not necessary" at this time doesn't sounds right, since we are agreed to go with this approach for long time ago.

I'd agree that we need to do refactoring since following the best practices of [navigation flow](https://developer.android.com/guide/navigation/use-graph/navigate#events) management, the screen level should expose the callback events. However, with this approach, we need to handle with UI/Integration Test to cover the logic.

**Reply by kaungkhantsoe** [reply]
@hoangnguyen92dn Indeed. We will need to do an integration test for that.
